### PR TITLE
Cleaned up the add + TLS example.

### PR DESCRIPTION
**Pull request by spikebike**
I added "example/add", cleaned it up, and added a readme.

It does run as is on my system.  I think it's useful as is, but am interested in any improvements/suggestions you have.

Nothing changed outside of example/add.

### CODE REVIEW

File: examples/add/README [hunk 1]
```diff
@@ -0,0 +1,21 @@
+
+This is a simple add service that happens over a TLS connection.
+
+To generate code from the protobuf file:
+cd addservice
+protoc --go_out=. addservice.proto
```
Comment (kylelemons) [inline examples/add/README]: In order to make the package as eminently goinstallable as possible, all of the generated files are actually checked into git.  You should be able to run the test.sh script to generate them the way it wants to.

File: examples/add/README [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+protoc --go_out=. addservice.proto 
+
+To compile the server:
+go build add.go
+
+To compile the client:
+go build client.go
+
+To make test certificates:
+./makecert.sh AnyFakeOrRealEmail@foo.com
```
Comment (kylelemons) [inline examples/add/README]: Since it's a throwaway, I would also generate (really small) certs with some email like dummy@example.com and check them in so that the demos just work.

File: examples/add/README [hunk 3]
```diff
@@ -0,0 +1,21 @@
+
+This is a simple add service that happens over a TLS connection.
+
+To generate code from the protobuf file:
+cd addservice
+protoc --go_out=. addservice.proto 
+
+To compile the server:
+go build add.go
```
Comment (kylelemons) [inline examples/add/README]: I would go with "go run add.go" and "go run client.go" (though it would make more sense if one had the word "server" in the name of the file.  I think I use "client" and "server" in the other examples) if you do it this way, though the other examples put them in subdirectories (which you might want to emulate for consistency).

File: examples/add/add.go [hunk 4]
```diff
@@ -0,0 +1,84 @@
+package main
+
+import (
+	"flag"
+	"log"
+	"net"
+	"crypto/rand"
+	"crypto/tls"
+	"crypto/x509"
+	"./addservice"
```
Comment (kylelemons) [inline examples/add/add.go]: Please use the absolute path and offset it by a blank line

File: examples/add/add.go [hunk 5]
```diff
@@ -0,0 +1,84 @@
+package main
```
Comment (kylelemons) [inline examples/add/add.go:1]: As I mentioned above, I would put this as examples/add/server/main.go

File: examples/add/add.go [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+
+// Add is the function that can be called remotely.  Note that this can be
+// called concurrently, so if the Echo structure did have internal state,
+// it should be designed for concurrent access.
+func (Add) Add(in *addservice.AddMessage, out *addservice.SumMessage) error {
+	log.Printf("server: X=%d", *in.X)
+	log.Printf("server: Y=%d", *in.Y)
+	out.Z = new(int32)
+	*out.Z = *in.X + *in.Y
+	log.Printf("server: Z=%d", *out.Z)
```
Comment (kylelemons) [inline examples/add/add.go]: I'd combine them into a single printf at the end; concurrent requests would get mixed up this way and be difficult later to track down if for some reason a request returned the wrong answer.

File: examples/add/add.go [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+	if ok {
+		log.Print("server: conn: type assert to TLS succeedded")
+		err := tlscon.Handshake()
+		if err != nil {
+			log.Fatalf("server: handshake failed: %s", err)
+		} else {
+			log.Print("server: conn: Handshake completed")
+		}
+		state := tlscon.ConnectionState()
+// Note we could reject clients if we don't like their public key.
```
Comment (kylelemons) [inline examples/add/add.go]: gofmt please.  This should also have a blank line above it.

File: examples/add/add.go [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+		if err != nil {
+			log.Fatalf("server: handshake failed: %s", err)
+		} else {
+			log.Print("server: conn: Handshake completed")
+		}
+		state := tlscon.ConnectionState()
+// Note we could reject clients if we don't like their public key.		
+		log.Println("Server: client public key is:")
+		for _, v := range state.PeerCertificates {
+			log.Print(x509.MarshalPKIXPublicKey(v.PublicKey))
```
Comment (kylelemons) [inline examples/add/add.go]: I'd remove 42 and replace this line with

log.Printf("server: client cert chain: %#v", v.Subject)

which seems potentially more interesting than a bunch of binary.  If you want to keep the PK bytes, make it log.Printf("%x") to compress them and make it more friendly (but it's pretty long).

File: examples/add/add.go [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+		}
+// Now that we have completed SSL/TLS 
+		addservice.ServeAddService(tlscon,Add{})
+	}
+}
+
+func serverTLSListen(service string) {
+
```
Comment (kylelemons) [inline examples/add/add.go:55]: dd

File: examples/add/add.go [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+// Load x509 certificates for our private/public key, makecert.sh will
+// generate them for you.
+
```
Comment (kylelemons) [inline examples/add/add.go:58]: dd

File: examples/add/add.go [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+			log.Print(x509.MarshalPKIXPublicKey(v.PublicKey))
+		}
+// Now that we have completed SSL/TLS 
+		addservice.ServeAddService(tlscon,Add{})
+	}
+}
+
+func serverTLSListen(service string) {
+
+// Load x509 certificates for our private/public key, makecert.sh will
```
Comment (kylelemons) [inline examples/add/add.go]: The note about makecert should go in a file comment at the top.

File: examples/add/add.go [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+// generate them for you.
+
+	cert, err := tls.LoadX509KeyPair("certs/server.pem", "certs/server.key")
+	if err != nil {
+		log.Fatalf("server: loadkeys: %s", err)
+	}
+// Note if we don't tls.RequireAnyClientCert client side certs are ignored.
```
Comment (kylelemons) [inline examples/add/add.go]: gofmt will align this, and put whitespace above it.  Same for most of the documentation comment that offset a separate "chunk" of code.

File: examples/add/add.go [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+	config := tls.Config{Certificates: []tls.Certificate{cert}, ClientAuth: tls.RequireAnyClientCert}
```
Comment (kylelemons) [inline examples/add/add.go]:
```
config := &tls.Config{
   Certificates: []tls.Certificate{cert},
   ClientAuth: tls.RequireAnyClientCert,
   Rand: rand.Reader, // only specify if this is not the default
}
```

File: examples/add/add.go [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+	config.Rand = rand.Reader
+	listener, err := tls.Listen("tcp", service, &config)
+	if err != nil {
+		log.Fatalf("server: listen: %s", err)
+	}
+	log.Print("server: listening")
```
Comment (kylelemons) [inline examples/add/add.go:74]: log.Printf("server: listening on %q", service)

File: examples/add/add.go [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+// Keep this loop simple/fast as to be able to handle new connections
+	for {
+		conn, err := listener.Accept()
+		if err != nil {
+			log.Printf("server: accept: %s", err)
+			break
+		}
+		log.Printf("server: accepted from %s", conn.RemoteAddr())
```
Comment (kylelemons) [inline examples/add/add.go:82]: server: new client %q

File: examples/add/README (hunk 1 above)
Comment (spikebike) [inline examples/add/README]: addservice.pb.go added to git.

test.sh does indeed generate addservice.pb.go.

Done.

File: examples/add/README (hunk 3 above)
Comment (spikebike) [inline examples/add/README]: I deleted the how to build proto, seems unecessary with the protobuf.go code checked in.

I renamed them to add-server and add-client.

Well, both example/echo and example/remote:
- Name the server echo or remote (no server in the name)
- Have the server in the basedir
- Have the client in a subdir

So that's what I did.

So seems like either server and client should be in subdirs, or neither.  For browsing/simplicity seems best to have all 3 go files in the basedir.

File: examples/add/README (hunk 2 above)
Comment (spikebike) [inline examples/add/README]: Done.

File: examples/add/add.go (hunk 5 above)
Comment (spikebike) [inline examples/add/add.go:1]: Plan to do that to echo/remote directories?  I change mine to whereever you want to go with the other 2.  For now it's add-server.go and add-client.go in the basedir.

File: examples/add/add.go (hunk 4 above)
Comment (spikebike) [inline examples/add/add.go]: Fixed absolute.  Should I override gofmt and add a blank line then the addservice line?

Oh, I don't think I can do it.  Hrm, if I make it:

```
"github.com/kylelemons/go-rpcgen/examples/add/addservice"
```

Then I can't compile it, because mine is a fork.  I'll leave it at   "github.com/kylelemons/go-rpcgen/examples/add/addservice" for now.

File: examples/add/add.go (hunk 6 above)
Comment (spikebike) [inline examples/add/add.go]: Done.

File: examples/add/add.go (hunk 7 above)
Comment (spikebike) [inline examples/add/add.go]: I did gofmt.  I just did ran gofmt again and it stays the same.  Did you mean above line 41 (it's not clear from the GUI which line you mean).

File: examples/add/add.go (hunk 8 above)
Comment (spikebike) [inline examples/add/add.go]: Well line 43 defines v, it is legal to have more than one cert, thus the for loop.  The contents of the public key are important for my use, but agree that it's not important for someone learning protobuf code.

The v.Subject is substantially longer and more ugly than the public key:
Key:
2012/04/05 23:02:00 Server: public key &{b0e5d7f400cd2ddb84c737b14990322cd5db3a1a14d1d8d767182c55134ac19d899ad5bd25fa1338b4edea37b60b867a8a44ea94ffd65efdc48cbc7c499fc71223c8fcaefbe66b4b4126d0080412805ab69b49ed2c496f373f2b7f1f27166bdb68adbd94f9f075a67969e148cd04ab6c2d581f2b975541ed58a5384bcbe2b3c5 10001}

2012/04/05 23:02:00 Server: cient cert chain pkix.Name{Country:[]string{"DE"}, Organization:[]string{"Random Company"}, OrganizationalUnit:[]string{"IT"}, Locality:[]string{"Earth"}, Province:[]string{"NRW"}, StreetAddress:[]string(nil), PostalCode:[]string(nil), SerialNumber:"", CommonName:"www.random.com", Names:[]pkix.AttributeTypeAndValue{pkix.AttributeTypeAndValue{Type:asn1.ObjectIdentifier{2, 5, 4, 6}, Value:"DE"}, pkix.AttributeTypeAndValue{Type:asn1.ObjectIdentifier{2, 5, 4, 8}, Value:"NRW"}, pkix.AttributeTypeAndValue{Type:asn1.ObjectIdentifier{2, 5, 4, 7}, Value:"Earth"}, pkix.AttributeTypeAndValue{Type:asn1.ObjectIdentifier{2, 5, 4, 10}, Value:"Random Company"}, pkix.AttributeTypeAndValue{Type:asn1.ObjectIdentifier{2, 5, 4, 11}, Value:"IT"}, pkix.AttributeTypeAndValue{Type:asn1.ObjectIdentifier{2, 5, 4, 3}, Value:"www.random.com"}, pkix.AttributeTypeAndValue{Type:asn1.ObjectIdentifier{1, 2, 840, 113549, 1, 9, 1}, Value:"dummy@example.com"}}}

It would be nice to print out the "dummy@example.com", but I couldn't figure out how to get to it.

File: examples/add/add.go (hunk 9 above)
Comment (spikebike) [inline examples/add/add.go:55]: Sorry, I see 8 lines of code, what does dd mean?

File: examples/add/add.go (hunk 11 above)
Comment (spikebike) [inline examples/add/add.go]: What file comment at the top?  The first comment in the file is about the Add struct{}

File: examples/add/add.go (hunk 12 above)
Comment (spikebike) [inline examples/add/add.go]: Do you use some flags with gofmt?  I can get the // to indent, but not add a blank space, I assume you mean above line 60.  Done.

The general rule is blank lines above all comments?

File: examples/add/add.go (hunk 13 above)
Comment (spikebike) [inline examples/add/add.go]: Heh, I figured gofmt would fix it, guess not.  Fixed.

File: examples/add/add.go (hunk 4 above)
Comment (kylelemons) [inline examples/add/add.go]: Move your fork over top of where my repository would normally be so that the paths line up.

File: example_ae/whoami/whoami.pb.go [hunk 16]
```diff
@@ -1,6 +1,4 @@
-// +build !appengine
```
Comment (kylelemons) [inline example_ae/whoami/whoami.pb.go:1]: Hmm, something is off here.  Did test.sh do this?  If so, please revert it manually and file a bug to fix test.sh.

File: examples/add/README [hunk 17]
```diff
@@ -0,0 +1,17 @@
+
```
Comment (kylelemons) [inline examples/add/README:1]: Can you markdown-format this file?

```
Summary
=======

This is a simple...

To run...
    go run...
```

etc

File: examples/add/certs/client.key [hunk 18]
```diff
@@ -0,0 +1,16 @@
+-----BEGIN PRIVATE KEY-----
```
Comment (kylelemons) [inline examples/add/certs/client.key:1]: Can you put a README in this directory saying that they're really insecure, throwaway certs and not to re-use them?

File: examples/add/example.output [hunk 19]
```diff
@@ -0,0 +1,38 @@
+
```
Comment (kylelemons) [inline examples/add/example.output]: I'd combine this with the README.md (markdown/github will format it nicely and show it in the directory listing online)

File: examples/remote/offload/offload.pb.go [hunk 20]
```diff
@@ -19,7 +19,7 @@ var _ = math.Inf
 
 type DataSet struct {
 	Data             *string `protobuf:"bytes,1,opt,name=data" json:"data,omitempty"`
-	XXX_unrecognized []byte  `json:",omitempty"`
+	XXX_unrecognized []byte  `json:"-"`
```
Comment (kylelemons) [inline examples/remote/offload/offload.pb.go:22]: My protoc-gen-go might be outdated (if so, you can file a bug against me to update), but these changes probably shouldn't be included in this pull request just to minimize the number of file touches.

**Comment by kylelemons** [comment]
It looks like you also missed adding the other files? I don't see the actual client/server anymore...

**Comment by spikebike** [comment]
Re-organized code, files, readme, printing, and related.  Closing this out because it's unwieldy, and will start a new one.
